### PR TITLE
Skip specs failing on TruffleRuby

### PR DESCRIPTION
**Pull request by andrykonchin**
Some pattern matching specs are failing on the recent TruffleRuby release 24.0. We will investigate the issues - some of them are known. So the failing specs are temporary disabled on TruffleRuby.

### CODE REVIEW

File: spec/language/ruby_pattern_matching_spec.rb [hunk 1]
```diff
@@ -1281,9 +1283,9 @@ def test_deconstruct_keys
...
 
     # https://github.com/jruby/jruby/issues/7854
-    unless defined?(JRUBY_VERSION)
+    unless defined?(JRUBY_VERSION) || RUBY_ENGINE == "truffleruby"
     assert_block do
-      case {}
+      case C.new({})
       in {}
         C.keys == nil
       end
```
Comment (andrykonchin) [inline spec/language/ruby_pattern_matching_spec.rb:1291]: I suppose it's a bug in https://github.com/ruby/ruby/blob/master/test/ruby/test_pattern_matching.rb that was hidden - `C.keys` returned keys captured in some previous test case, that by chance matches `C.new({})` and captures the `nil` value.

Comment (palkan) [reply spec/language/ruby_pattern_matching_spec.rb:1291]: Good catch! Yeah, having `case {} in ` and checking for `C.keys` doesn't make any sense here

**Review by palkan — COMMENTED** [review]
Thanks!

One thing: could you please add links to the corresponding TruffleRuby repo issues where we skip it? (If any)

I check them from time to time to see if we can unskip the tests back.

**Comment by andrykonchin** [comment]
There are no issues for not supported pattern matching edge cases. But failing ruby/spec and MRI tests are tracked in plain text files:
- https://github.com/oracle/truffleruby/blob/master/spec/tags/language/pattern_matching_tags.txt
- https://github.com/oracle/truffleruby/blob/master/test/mri/excludes/TestPatternMatching.rb

Would you like to have these links added (for every skipped test case)?

File: spec/language/pattern_matching_spec.rb [hunk 2]
```diff
@@ -722,6 +725,8 @@ def obj.deconstruct; "" end
       end
 
       it "accepts a subclass of Array from #deconstruct" do
+        next skip if RUBY_ENGINE == "truffleruby"
+
```
Comment (andrykonchin) [inline spec/language/pattern_matching_spec.rb:729]: BTW this test case differs a bit from the one in the `ruby/spec` repository. Not sure if it's done intentionally. Here the test relies on a way how CRuby accesses array elements - with `#[]` method. The `ruby/spec`'s current version doesn't redefine the `#[]` method:

```ruby
    it "accepts a subclass of Array from #deconstruct" do
      obj = Object.new
      def obj.deconstruct
        Class.new(Array).new([0, 1])
      end

      eval(<<~RUBY).should == true
        case obj
        in [1, 2]
          false
        in [0, 1]
          true
        end
      RUBY
    end
```

This version passes on TruffleRuby but the version with overridden `#[]` - doesn't.

**Comment by palkan** [comment]
> Would you like to have these links added (for every skipped test case)?

Let's add the links to the top of the spec file (`ruby_pattern_matching_spec.rb`).

File: spec/language/pattern_matching_spec.rb (hunk 2 above)
Comment (palkan) [reply spec/language/pattern_matching_spec.rb:729]: Looks like the change to `ruby/spec` was made for TruffleRuby: https://github.com/ruby/spec/commit/0e042687332c12a9f0468b52c1dc12d4a5a7e0fa

Unfortunately, there are no additional details on that change.

MRI uses `#[]` (or `opt_aref`) to access elements returned by `#deconstruct`. So, you can do smth like this:

```ruby
class A < Array
  def [](i) = at(-i-1)
end

a = A.new([1, 2])

case a
in [2, 1]
  true
end #=> true
```

I don't know if it's an intended feature or not; @k-tsj could you please take a look?

File: spec/language/ruby_pattern_matching_spec.rb (hunk 1 above)
Comment (andrykonchin) [reply spec/language/ruby_pattern_matching_spec.rb:1294]: The fix in the Ruby repo - https://github.com/ruby/ruby/pull/10757

**Comment by andrykonchin** [comment]
Done

File: spec/language/pattern_matching_spec.rb (hunk 2 above)
Comment (k-tsj) [reply spec/language/pattern_matching_spec.rb:732]: @palkan 
Yes, it's intentional. If `#[]` is redefined, it should be called,.

Comment (palkan) [reply spec/language/pattern_matching_spec.rb:732]: @k-tsj Thank you!

@eregon ^ That's what I mentioned yesterday.

Comment (eregon) [reply spec/language/pattern_matching_spec.rb:732]: I see. I'm not sure it's a good idea to have that behavior in Ruby, maybe something to discuss on ruby-core.
Subclassing Array and redefining `[]` on its own is already clearly a bad idea, so it seems a niche edge case no one should use.

**Comment by palkan** [comment]
@andrykonchin Sorry, missed latest updates. Looks good, thanks!
